src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -77,31 +79,42 @@ private sealed class LatestDiagnosticsForSpanGetter
                  Func<string, IDisposable?>? addOperationScope,
                  bool includeSuppressedDiagnostics,
                  CodeActionRequestPriority priority,
-                 string? diagnosticId,
+                 Func<string, bool>? shouldIncludeDiagnostic,
                  CancellationToken cancellationToken)
             {
                 var stateSets = owner._stateManager
                                      .GetOrCreateStateSets(document.Project).Where(s => !owner.DiagnosticAnalyzerInfoCache.IsAnalyzerSuppressed(s.Analyzer, document.Project));
 
-                // filter to specific diagnostic it is looking for
-                if (diagnosticId != null)
-                {
-                    stateSets = stateSets.Where(s => owner.DiagnosticAnalyzerInfoCache.GetDiagnosticDescriptors(s.Analyzer).Any(d => d.Id == diagnosticId)).ToList();
-                }
-
-                var compilationWithAnalyzers = await CreateCompilationWithAnalyzersAsync(document.Project, stateSets, includeSuppressedDiagnostics, cancellationToken).ConfigureAwait(false);
+                var compilationWithAnalyzers = await GetOrCreateCompilationWithAnalyzersAsync(document.Project, stateSets, includeSuppressedDiagnostics, cancellationToken).ConfigureAwait(false);
 
                 return new LatestDiagnosticsForSpanGetter(
-                    owner, compilationWithAnalyzers, document, stateSets, diagnosticId, range,
+                    owner, compilationWithAnalyzers, document, stateSets, shouldIncludeDiagnostic, range,
                     blockForData, addOperationScope, includeSuppressedDiagnostics, priority);
             }
 
+            private static async Task<CompilationWithAnalyzers?> GetOrCreateCompilationWithAnalyzersAsync(
+                Project project,
+                IEnumerable<StateSet> stateSets,
+                bool includeSuppressedDiagnostics,
+                CancellationToken cancellationToken)
+            {
+                if (_lastProjectAndCompilationsWithAnalyzers.TryGetTarget(out var projectAndCompilationWithAnalyzers) &&
+                    projectAndCompilationWithAnalyzers?.Project == project)
+                {
+                    return projectAndCompilationWithAnalyzers.CompilationWithAnalyzers;
```

Comment:
does this really need to be a weak ref?  we're going to be calling this non-stop right?  so it's not awful to just make it a strong ref.

Reply:
I feel hesitant to have a static strong reference to Project/Compilation.